### PR TITLE
switch the Matlab RPC tool from LCM to Protobuf + File I/O

### PR DESCRIPTION
**Pull request by RussTedrake**
Due to limitations in LCM + Matlab (https://github.com/lcm-proj/lcm/issues/147) and limitations observed where matlab could not keep up with the c++ publishing ... and would crash.

Now the RPC calls are written to a protobuf and dumped into a file.  Matlab can load the file and run the commands (multiple times, if desired).  Synchronous plotting is still available if you just make a named pipe.

@hongkai-dai -- this means we can use new versions of matlab on linux.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/5786)

### CODE REVIEW

**Comment by hongkai-dai** [comment]
Awesome! Thanks for working on this. The MATLAB R2014a does not support saving to svg files, which makes visualization harder. The newer version of MATLAB supports svg format. Kudos for getting this working!

**Comment by RussTedrake** [comment]
Note:  To get things work on MATLAB R2017a on linux, I had to point the matlab `libstdc++` to system `libstdc++`:
```
sudo rm /usr/local/MATLAB/R2017a/sys/os/glnxa64/libstdc++.so.6
sudo ln -s /usr/lib/gcc/x86_64-linux-gnu/5.4.0/libstdc++.so /usr/local/MATLAB/R2017a/sys/os/glnxa64/libstdc++.so.6
```

File: drake/common/CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,12 @@
 # Drake common library
 #
 
+drake_find_package(Protobuf REQUIRED)
+include_directories(${Protobuf_INCLUDE_DIRS})
+include_directories(${PROJECT_BINARY_DIR}/..)  # to find generated protobuf headers
```
Comment (RussTedrake) [inline drake/common/CMakeLists.txt]: @jamiesnape -- i wonder if you could take a look at this for me?  it works on my system and half the build servers, but is not robust enough for the entire build matrix.

Comment (jamiesnape) [inline drake/common/CMakeLists.txt]: @RussTedrake Check the last three commits I added to this branch: https://github.com/jamiesnape/drake/tree/call_matlab_protobuf

**Comment by RussTedrake** [comment]
@jamiesnape -- thanks!!  (bummer that we have to include our own FindProtobuf.cmake).  I'll merge your changes in here.

**Comment by jamiesnape** [comment]
> (bummer that we have to include our own FindProtobuf.cmake)

Yes, they only fixed that issue in CMake 3.8 (which was released just a few hours ago).

**Comment by RussTedrake** [comment]
+@hongkai-dai for feature review.  could you give it a quick spin, please?

---

Review status: 0 of 33 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-:-KhPmDFj75yeCFPVPYyZ:b-fxeob4)*

**Comment by RussTedrake** [comment]
note that the changes are not quite as severe as suggested by the github/reviewable displays, as many of the files MOVED... (and updated) but appear here as deleted and brand new files reappearing.  

---

Review status: 0 of 33 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-:-KhPmOG7VpIuLbK8RNPs:b-1zndx5)*

**Comment by hongkai-dai** [comment]
Tested this on my laptop, it works great! Thanks!

---

Reviewed 30 of 32 files at r1, 3 of 5 files at r2.
Review status: 31 of 33 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/matlab/util/call_matlab_client.cc, line 76 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhPzxJQRD4JONr7sFyz:-KhPzxJREW7OXrrd-nFo:bl65ltg) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/matlab/util/call_matlab_client.cc#L76)):*
> ```cpp
>         for (int j = 0; j < i; j++) {
>           mxDestroyArray(rhs[j]);
>         }
> ```

Should we early exit here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-:-KhTJ2cPvJtSAcfbyXI5:b-rt0yey)*

**Comment by hongkai-dai** [comment]
:lgtm: +@sherm1 for platform review please

---

Review status: 31 of 33 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-:-KhTJ8h1t7adjE44Ptrl:bq0dloy)*

**Comment by sherm1** [comment]
Platform: some BTWs. Mainly it isn't clear to me that the doxygen ends up in a useful place -- ptal. Also please consider whether this might need a little .rst developer documentation so people know this is available. Otherwise :lgtm:.

The CMake dependence makes me nervous but I don't have a suggestion to improve it -- will this still require CMake after we switch entirely to Bazel?

---

Reviewed 30 of 32 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/common/call_matlab.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhTU8K7nkNKN-fwIEIs:-KhTU8K7nkNKN-fwIEIt:bmntiuj) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/common/call_matlab.h#L30)):*
> ```objc
> /// reference" that is kept by the publisher.
> ///
> /// See call_matlab_test.cc for some simple examples.
> ```

BTW I believe this doxygen comment attaches to `namespace drake` since that is what immediately follows. In that case I don't think anyone will see it, except possibly by accident. Please consider creating a "Call Matlab" module or attaching it to some Matlab-related class.

---

*[drake/common/call_matlab.h, line 37 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhTV5lohSm7Ntbys2X6:-KhTV5lohSm7Ntbys2X7:b-d721j5) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/common/call_matlab.h#L37)):*
> ```objc
> /// Serialize our favorite data types into the matlab_array structure.
> /// To support a calling call_matlab for a new data type, simply implement
> //  another one of these methods.
> ```

BTW mixed "///" and "//" and also not clear where this doxygen will come out. Probably should just be "//".

---

*[drake/common/matlab_rpc.proto, line 13 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhTRo9IXncA7iQpgJtO:-KhTRo9JBzo6GXMtnGsk:b-rss5y4) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/common/matlab_rpc.proto#L13)):*
> ```protobuf
> {
>   enum ArrayType {
>     REMOTE_VARIABLE_REFERENCE = 0; // encoded as an int64 used in the lhs argument of lcmt_matlab_feval; rows and cols must be set to 1.
> ```

BTW wrap at 80 characters if possible here (and below).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-:-KhTTj_39STXL1CVckaq:b-qobonf)*

**Comment by RussTedrake** [comment]
Adding .rst in the future sounds good.  And no -- the soon-to-be-available solution for cmake is to have cmake build the mex, but link against the drakeCommon built by bazel.  

---

Review status: 31 of 35 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/common/call_matlab.h, line 30 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhTU8K7nkNKN-fwIEIs:-KhZoB6-3cd9_UaGURxN:b-shim6k) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/common/call_matlab.h#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I believe this doxygen comment attaches to `namespace drake` since that is what immediately follows. In that case I don't think anyone will see it, except possibly by accident. Please consider creating a "Call Matlab" module or attaching it to some Matlab-related class.
</blockquote></details>

used the @file tag.  still not easy to find in doxygen, but seems consistent with many other uses in drake.

---

*[drake/common/call_matlab.h, line 37 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhTV5lohSm7Ntbys2X6:-KhZo9R53_hgN3BJ32WP:b-qssebl) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/common/call_matlab.h#L37)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW mixed "///" and "//" and also not clear where this doxygen will come out. Probably should just be "//".
</blockquote></details>

fixed.

---

*[drake/matlab/util/call_matlab_client.cc, line 76 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-KhPzxJQRD4JONr7sFyz:-KhZpagQgQpduvaPWpD_:b-1ij47n) ([raw file](https://github.com/robotlocomotion/drake/blob/c20f7db837ad1625189fb285188542d17b119969/drake/matlab/util/call_matlab_client.cc#L76)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Should we early exit here?
</blockquote></details>

this was really debugging code and does not belong.  zapped it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-:-KhZM0hlO5lDU866qD6p:b-u1oo81)*

**Comment by sherm1** [comment]
Reviewed 1 of 32 files at r1, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/matlab_rpc.proto, line 43 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-Kh_1avoNpfRcRSMVh93:-Kh_1avoNpfRcRSMVh94:b-tw390l) ([raw file](https://github.com/robotlocomotion/drake/blob/1af1fdb554a093165c96ac04716552a95570d379/drake/common/matlab_rpc.proto#L43)):*
> ```protobuf
> message MatlabRPC {
>   repeated int64 lhs = 1;  // unique id for variables held in the MATLAB
>   	   	       	   // client's workspace
> ```

BTW this has tabs. Probably should be only spaces in .proto as elsewhere in Drake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786)*

**Comment by RussTedrake** [comment]
Review status: 32 of 34 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/matlab_rpc.proto, line 43 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5786#-Kh_1avoNpfRcRSMVh93:-KhaphVcvzzW8Mm2w5D3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/1af1fdb554a093165c96ac04716552a95570d379/drake/common/matlab_rpc.proto#L43)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this has tabs. Probably should be only spaces in .proto as elsewhere in Drake.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5786)*

**Comment by jwnimmer-tri** [comment]
FYI Calling `find_package` directly in `drake/common/CMakeLists.txt` was wrong.  (It's always wrong to bury calls like that within folders.)  See #6001.

**Pull request merged**